### ISSUE TITLE
Potential memory leak in local.py

### ISSUE THREAD

**Issue by mariusmagureanu** [pallets/werkzeug#2205]
**The problem**:

The memory footprint of a flask application keeps increasing as requests are coming in, suggesting a leak.

**Some context**:

The application is a rather simple one, the testing has been done against a simple endpoint returning an empty string, so no heavy work or anything else is done under the hood.

The application uses ``gevent`` and thus ``greenlet`` as well.
While profiling the app it turns out that ``greenlet`` objects are not collected.

This behavior is encountered when using ``Werkzeug 2.0.1``.
However, if downgrading to ``Werkzeug 1.0.1`` the memory usage is rather constant and the application works as expected.

I wonder if [this](https://github.com/pallets/werkzeug/blob/main/src/werkzeug/local.py#L48) may be related in any way, as the older implementation in Werkzeug 1.x is working as expected.

**Some more context**:

This is a graph based on the application running with Werkzeug 1.0.1

![greenlets_w1_loop](https://user-images.githubusercontent.com/2333090/128420075-ab54e8e4-4d70-4ee4-b589-aec9b6a8325b.png)

This is a graph based on the same application running with Werkzeug 2.0.1

![greenlets1](https://user-images.githubusercontent.com/2333090/128420299-93a6844b-16d0-4ca3-ba1c-606e682302be.png)

In both situations the application has been fed with 10k reqs.

Environment:

- Python version: 3.7
- Werkzeug version: 2.0.1
- Gevent 1.3.7
- Greenlet 1.0.1

Any help or ideas in this regards are much appreciated.

**Comment by davidism** [comment]
If greenlet isn't collected, then that sounds like a bug with greenlet.

**Comment by mariusmagureanu** [comment]
Right, that was my first thought as well, but in such case I'd expect to have the same leak with Werkzeug 1.0.1 as well. Maybe ContextVar doesn't "play" that well with Greenlet?

**Comment by davidism** [comment]
I barely know anything about greenlet or ConextVar, let alone the internal interactions with the gc. You're welcome to look into this more, but I won't be able to. Again, "greenlet doesn't play well with contextvar" sounds like a bug in greenlet.

**Comment by mariusmagureanu** [comment]
Sure thing, this defenitelly requires more looking into. I was merely curious if such a behavior has been met before. 
The only thing that puzzled me is that 1.0.1 is working fine whereas 2.0.1 does not.
Then again, the usage of ContextVar has been introduced in 2.0.0; here #1778 

Regardless, thank you for your input.

**Comment by davidism** [comment]
@pgjones any thoughts here?

**Comment by mariusmagureanu** [comment]
One thing I noticed in v1.0 is [this](https://github.com/pallets/werkzeug/blob/54acdd16b247f7037482737e72ec52fc6d50a78d/src/werkzeug/local.py#L67), essentially the greenlet object is just removed from the storage dict.

In v2.0, the ``__release_local__()`` call ends up [here](https://github.com/pallets/werkzeug/blob/main/src/werkzeug/local.py#L79) - where the greenlet objects keep being added in the storage.

Simply updating the ``__release_local__()`` method in v2 to do the same removal as in v1 seems to lead to a correct behavior.

However, I'm not insisting that this is the solution  as I'm not familiar with neither the werkzeug or greenlet code base.

But I'd gladly assist in debugging, if required.

**Comment by pgjones** [comment]
@mariusmagureanu "Simply updating the __release_local__() method in v2" - could you share the diff? I think with the greenlet version you are using (v 1.0.1) `greenlet.GREENLET_USE_CONTEXT_VARS` will be `True` and a stdlib ContextVar will be used, rather than the Fake one you've linked to so I'd like to understand how this helps.

**Comment by mariusmagureanu** [comment]
Here comes the diff:

```
diff --git a/src/werkzeug/local.py b/src/werkzeug/local.py
index a5a7870e..fbbe63af 100644
--- a/src/werkzeug/local.py
+++ b/src/werkzeug/local.py
@@ -145,7 +145,7 @@ class Local:
         return LocalProxy(self, proxy)
 
     def __release_local__(self) -> None:
-        self._storage.set({})
+        self._storage.storage.pop(_get_ident(), None)
 
     def __getattr__(self, name: str) -> t.Any:
         values = self._storage.get({})
```

I can't totally rely on the stdlib ContextVar usage, my app may also run on python 3.6; If I'm not mistaking stdlib ContextVar has been introduced in 3.7;

**Comment by pgjones** [comment]
Thanks, that makes sense. I think I've fixed in #2212.

**Comment by mariusmagureanu** [comment]
@pgjones Thanks for your help.
